File: modlab/core/pack_manager.py
import json
from pathlib import Path

from PySide6.QtCore import QObject, Signal

from .pack import Loader, Pack, PackStatus, loader_from_text


class PackManager(QObject):
    """Owns the in-memory pack list and mirrors the backend packmapping file."""

    packs_changed = Signal()
    pack_progress_changed = Signal(str, int)

    def __init__(self, parent=None):
        super().__init__(parent)
        self._packs: list[Pack] = []
        self.mapping_path = Path.home() / ".modlab" / "packmapping" / "packmapping.json"

    @property
    def packs(self) -> list[Pack]:
        return self._packs

    def load(self) -> None:
        self.reload_from_disk()

    def reload_from_disk(self) -> None:
        self._packs.clear()
        if not self.mapping_path.exists():
            self.packs_changed.emit()
            return

        try:
            root = json.loads(self.mapping_path.read_text(encoding="utf-8") or "{}")
        except json.JSONDecodeError:
            self.packs_changed.emit()
            return

        if not isinstance(root, dict):
            self.packs_changed.emit()
            return

        for key in sorted(root):
            obj = root.get(key) or {}
            if not isinstance(obj, dict):
                continue
            self._packs.append(
                Pack(
                    key=key,
                    name=str(obj.get("name") or key),
                    mc_version=str(obj.get("version") or ""),
                    loader=loader_from_text(str(obj.get("loader") or "vanilla")),
                    mod_count=int(obj.get("mod_count") or 0),
                    status=PackStatus.READY
                    if "launch_version" in obj
                    else PackStatus.NOT_INSTALLED,
                )
            )
        self.packs_changed.emit()

    def save(self) -> None:
        if not self.mapping_path.exists():
            return
        try:
            root = json.loads(self.mapping_path.read_text(encoding="utf-8") or "{}")
        except json.JSONDecodeError:
            return
        if not isinstance(root, dict):
            return

        for pack in self._packs:
            if pack.key not in root or not isinstance(root[pack.key], dict):
                continue
            root[pack.key]["mod_count"] = pack.mod_count

        self.mapping_path.write_text(json.dumps(root, indent=4), encoding="utf-8")

    def find_pack(self, key: str) -> Pack | None:
        for pack in self._packs:
            if pack.key == key:
                return pack
        return None

    def set_progress(self, key: str, progress: int) -> None:
        pack = self.find_pack(key)
        if pack is None:
            return
        pack.install_progress = progress
        self.pack_progress_changed.emit(key, progress)

    def set_status(self, key: str, status: PackStatus) -> None:
        pack = self.find_pack(key)
        if pack is None:
            return
        pack.status = status
        self.packs_changed.emit()

File: modlab-cpp-client/modlab/src/core/PackManager.cpp
#include "PackManager.h"
#include <QDir>
#include <QFile>
#include <QJsonDocument>
#include <QJsonObject>
#include <QStandardPaths>

PackManager::PackManager(QObject* parent)
    : QObject(parent)
{
    // Mirror the Python backend's ~/.modlab-pyshit-client/packmapping/packmapping.json
    QString home = QDir::homePath();
    m_mappingPath = home + "/.modlab-pyshit-client/packmapping/packmapping.json";
}

void PackManager::load() {
    reloadFromDisk();
}

void PackManager::reloadFromDisk() {
    m_packs.clear();

    QFile f(m_mappingPath);
    if (!f.open(QIODevice::ReadOnly)) return;

    QJsonDocument doc = QJsonDocument::fromJson(f.readAll());
    f.close();

    if (!doc.isObject()) return;
    QJsonObject root = doc.object();

    for (const QString& key : root.keys()) {
        QJsonObject obj = root[key].toObject();

        Pack p;
        p.key       = key;
        p.name      = obj["name"].toString(key);
        p.mcVersion = obj["version"].toString();
        p.modCount  = obj["mod_count"].toInt(0);

        QString loaderStr = obj["loader"].toString("vanilla").toLower();
        if      (loaderStr == "fabric") p.loader = Loader::Fabric;
        else if (loaderStr == "forge")  p.loader = Loader::Forge;
        else                            p.loader = Loader::Vanilla;

        // If launch_version is set the pack has been installed at least once
        p.status = obj.contains("launch_version")
            ? PackStatus::Ready
            : PackStatus::NotInstalled;

        m_packs.append(p);
    }

    emit packsChanged();
}

void PackManager::save() {
    // Qt side only writes status/mod_count extras — never overwrites
    // the Python-owned fields. For safety, reload first then patch.
    QFile f(m_mappingPath);
    if (!f.open(QIODevice::ReadOnly)) return;
    QJsonDocument doc = QJsonDocument::fromJson(f.readAll());
    f.close();
    if (!doc.isObject()) return;

    QJsonObject root = doc.object();
    for (const Pack& p : m_packs) {
        if (!root.contains(p.key)) continue;
        QJsonObject obj = root[p.key].toObject();
        obj["mod_count"] = p.modCount;
        root[p.key] = obj;
    }

    if (!f.open(QIODevice::WriteOnly | QIODevice::Truncate)) return;
    f.write(QJsonDocument(root).toJson(QJsonDocument::Indented));
    f.close();
}

Pack* PackManager::findPack(const QString& key) {
    for (Pack& p : m_packs)
        if (p.key == key) return &p;
    return nullptr;
}

void PackManager::setProgress(const QString& key, int progress) {
    if (Pack* p = findPack(key)) {
        p->installProgress = progress;
        emit packProgressChanged(key, progress);
    }
}

void PackManager::setStatus(const QString& key, PackStatus status) {
    if (Pack* p = findPack(key)) {
        p->status = status;
        emit packsChanged();
    }
}
